### PR TITLE
Fix fromCIDR policy on kernels 4.10 or older and extend test coverage

### PR DESCRIPTION
**Pull request by willdeuschle**
Builds on the draft PR originally opened by @joestringer #10580

That PR body:

This series extends the runtime CI tests to validate that fromCIDR policy works correctly using a local docker network, similar to how existing FQDN tests do.

Review patch-by-patch, there is some large refactoring of the CIDR tests in policy but it's mostly kept to one non-functional patch.
On top of that, see the final commit I added. The tests were failing in CI (and for me locally) because for kernel versions without LPM maps, ipcache_lookup4 is unable to lookup CIDRs based on a prefix other than a /32. Changing from ipcache_lookup4 to lookup_ip4_remote_endpoint fixes this because the latter function uses unrolled hash map lookups to simulate a LPM map lookups for kernels that don't support LPM maps. See bpf/lib/eps.h for further details on that.

It may be desirable to replace usage of ipcache_lookup[46] everywhere with lookup_ip[46]_remote_endpoint but that is beyond the scope of this PR.

### CODE REVIEW

**Comment by nebril** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
BPF changes LGTM. I suggest someone else reviews my CI changes in this PR :-)

@Will would you mind renaming the PR to something like "Fix fromCIDR policy on kernels 4.10 or older and extend test coverage"? This will be useful for generating the release notes to clearly highlight which release gets this fix.

**Comment by will** [comment]
> BPF changes LGTM. I suggest someone else reviews my CI changes in this PR :-)
> 
> @will would you mind renaming the PR to something like "Fix fromCIDR policy on kernels 4.10 or older and extend test coverage"? This will be useful for generating the release notes to clearly highlight which release gets this fix.

Sorry, it doesn't look like I have permissions to change the name of the PR.

**Comment by willdeuschle** [comment]
Updated, looks like the test failure was an etcd flake

**Comment by joestringer** [comment]
@will ha! My apologies, my not-so-nimble fingers intended that for @willdeuschle ;-)

**Comment by joestringer** [comment]
@willdeuschle I wouldn't worry too much about the GKE CI run right now, it's not marked required and there are known flakes on that job at the moment. I think we just need CI review.

**Review by ianvernon — APPROVED** [review]
Nice work!

File: bpf/bpf_netdev.c [hunk 1]
```diff
@@ -270,7 +271,7 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, const struct iphdr *ip4,
 
 	/* Packets from the proxy will already have a real identity. */
 	if (identity_is_reserved(srcid_from_ipcache)) {
-		info = ipcache_lookup4(&IPCACHE_MAP, ip4->saddr, V4_CACHE_KEY_LEN);
+		info = lookup_ip4_remote_endpoint(ip4->saddr);
```
Comment (pchaigno) [inline bpf/bpf_netdev.c:274]: Don't we need the same fix for the IPv6 path (i.e., `resolve_srcid_ipv6`)?
/cc @joestringer

**Review by pchaigno — CHANGES_REQUESTED** [review]

**Comment by pchaigno** [comment]
It looks like the bug was [introduced in 1.2](https://github.com/cilium/cilium/commit/bfdfc3dea6b78b5af3d4d94d435ae9ad6ea26aa1) so we would at least need to backport to 1.7. Do we also want to backport to previous releases? As far as I understand the impact of the bug is unexpected packet drops, right?
For the 1.7 backport, do we usually backport tests alongside the fix or should we just backport that last commit?

File: bpf/bpf_netdev.c (hunk 1 above)
Comment (willdeuschle) [reply bpf/bpf_netdev.c:274]: I think yes - but the implementation of `ipcache_lookup6` is slightly different and I've only run tests with ipv4 packets

**Comment by joestringer** [comment]
I'm not sure whether the tests will backport cleanly as they might be relying on other refactoring that occurred during the v1.8 dev cycle. I don't mind taking on the backport of the functional commit only to v1.7. I'd probably avoid backporting further just given that we have no user reports of this issue.

File: bpf/bpf_netdev.c (hunk 1 above)
Comment (joestringer) [reply bpf/bpf_netdev.c:274]: I think it wouldn't hurt to make the equivalent change there, as long as it passes CI.

We can separately review how well we cover IPv6 in each of the CI tests but I suspect that the scope of that change is bigger than what we intend to address in this PR.

Comment (willdeuschle) [reply bpf/bpf_netdev.c:274]: added the ipv6 change as well, would be good to run CI again

Comment (pchaigno) [reply bpf/bpf_netdev.c:274]: @willdeuschle I'm not seeing it. Maybe you pushed the wrong commit?

Comment (willdeuschle) [reply bpf/bpf_netdev.c:274]: apologies - wrong remote. it's fixed now

**Comment by joestringer** [comment]
test-me-please

**Review by pchaigno — APPROVED** [review]

**Pull request merged**